onecodex/lib/upload.py

```diff
@@ -366,10 +364,56 @@ def _upload_document_fileobj(file_obj, file_name, documents_resource):
         documents_resource._client._root_url + fields["callback_url"],  # full callback url
     )
 
-    msg = "{}: finished".format(file_name)
+    msg = f"{file_name}: finished"
     document_id = s3_upload.get("document_id")
     if document_id is not None:
-        msg += " as document {}".format(document_id)
+        msg += f" as document {document_id}"
+
+    log.info(msg)
+    return document_id
+
+
+def _upload_asset_fileobj(file_obj, file_name, assets_resource):
+    """Upload a single file-like object to a One Codex Asset directly to S3.
+
+    Parameters
+    ----------
+    file_obj : `FilePassthru`, or a file-like object
+        If a file-like object is given, its mime-type will be sent as 'text/plain'. Otherwise,
+        `FilePassthru` will send a compressed type if the file is gzip'd or bzip'd.
+    file_name : `string`
+        The file_name you wish to associate this file with at One Codex.
+    assets_resource : `onecodex.models.Assets`
+        Wrapped potion-client object exposing `init_upload` and `confirm_upload` routes to mainline.
+
+    Raises
+    ------
+    UploadException
+        In the case of a fatal exception during an upload.
+
+    Returns
+    -------
+    `string` id of newly uploaded asset.
+    """
+    try:
+        fields = assets_resource.init_multipart_upload({"name": file_name})
+    except requests.exceptions.HTTPError as e:
+        raise_api_error(e.response, state="init")
+    except requests.exceptions.ConnectionError:
+        raise_connectivity_error(file_name)
+
+    s3_upload = _s3_intermediate_upload(
+        file_obj,
+        file_name,
+        fields,
+        assets_resource._client.session,
+        assets_resource._client._root_url + fields["callback_url"],  # full callback url
+    )
+
+    msg = f"{file_name}: finished"
+    document_id = s3_upload.get("document_id")
+    if document_id is not None:
+        msg += f" as document {document_id}"
```

Comment:
Shouldn't it be `asset_id`?

Reply:
Ah, actually `asset_id` isn't in the response.